### PR TITLE
[DPE-4880] TLS CA rotation in large deployments

### PR DESCRIPTION
**Pull request by reneradoi**
This PR adds integration testing for TLS CA rotation in large deployments as well as the functionality to enable it, by setting the flags for ongoing TLS CA rotation not only to the `opensearch-peers` relation databag, but also to the relations `peer-cluster` and `peer-cluster-orchestrator`. By that, all related units in a large deployment broadcast their current TLS CA rotation status to the relation databags.

### CODE REVIEW

**Review by phvalguima — CHANGES_REQUESTED** [review]
Nice! I left a comment in the test coverage, besides checking the CA, I'd recommend trying to get the `ContinuousWrites` class to load the new CA as well and explicitly checking it.

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks René! I noticed a bit of an odd behavior in the CI run logs, that indicates that the a unit stayed in `Waiting for TLS to be configured...` for over 4min (20:32:28 - 20:36:39)
Any idea what that might be due to?

**Comment by reneradoi** [comment]
> Thanks René! I noticed a bit of an odd behavior in the CI run logs, that indicates that the a unit stayed in `Waiting for TLS to be configured...` for over 4min (20:32:28 - 20:36:39) Any idea what that might be due to?

Looks like this some odd log output from the `wait_until` method, it only logs the first application it can find, not all. According to the logs, at this time the `opensearch-data-0` unit was updated.

File: tests/integration/tls/test_ca_rotation.py [hunk 1]
```diff
@@ -99,3 +112,113 @@ async def test_rollout_new_ca(ops_test: OpsTest) -> None:
...
+            FAILOVER_APP: {"active": []},
+        },
+        units_statuses=["active"],
+        wait_for_exact_units={app: units for app, units in APP_UNITS.items()},
+        timeout=2400,
+        idle_period=IDLE_PERIOD,
+    )
+
+    more_writes = await c_writes.count()
+    await c_writes.stop()
```
[the remark this answers is not in the document]
Comment (reneradoi) [reply tests/integration/tls/test_ca_rotation.py:221]: added to the test

**Review by skourta — APPROVED** [review]

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 2]
```diff
@@ -1142,6 +1142,10 @@ def _post_start_init(self, event: _StartOpenSearch):  # noqa: C901
         if self.opensearch_peer_cm.is_provider():
             self.peer_cluster_provider.refresh_relation_data(event, can_defer=False)
 
+        # before resetting the CA rotation state, we remove the old ca from the truststore
+        if self.peers_data.get(Scope.UNIT, "tls_ca_renewed", False):
+            self.tls.remove_old_ca()
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1147]: This should only happen if all nodes / all applications have finished rotating the certificates after a CA rotation. I think we need an additional flag on the peer-cluster relation data if it's a large deployment, no?

File: lib/charms/opensearch/v0/opensearch_base_charm.py [hunk 3]
```diff
@@ -1142,6 +1142,10 @@ def _post_start_init(self, event: _StartOpenSearch):  # noqa: C901
         if self.opensearch_peer_cm.is_provider():
             self.peer_cluster_provider.refresh_relation_data(event, can_defer=False)
 
+        # before resetting the CA rotation state, we remove the old ca from the truststore
+        if self.peers_data.get(Scope.UNIT, "tls_ca_renewed", False):
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_base_charm.py:1146]: I think we should instead check on `ca_rotation_complete_in_cluster` to ensure that all nodes / all applications have finished rotating the certificates after a CA rotation, no?

**Review by Mehdi-Bendriss — APPROVED** [review]
Thanks René - approved with a question

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 2 above)
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1147]: Yes, currently the `old-ca` would be removed from the truststore even before all cluster-members have updated the new certs. But the truststore file would only be loaded on another restart, so it does not get active immediately.

File: lib/charms/opensearch/v0/opensearch_base_charm.py (hunk 3 above)
Comment (reneradoi) [reply lib/charms/opensearch/v0/opensearch_base_charm.py:1146]: But that would never happen at the same time, or we would need to defer the `_StartOpenSearch` event here.

**Pull request merged**